### PR TITLE
Fixing service-restart testcases.

### PR DESCRIPTION
**Pull request by rraghav-cisco**
### Description of PR

Summary:
The pfcwd_basic service-restart cases keep failing due to: https://github.com/sonic-net/sonic-buildimage/issues/20637

The ask is not to restart swss multiple times without doing a config reload in between. So in this PR:
1.  we are doing config-reload for every iteration of the test 
2. The swss restart is done only once in one DUT.
3. The asic is randomly picked, and the swss of that ASIC is restarted instead of doing the restart for all asics.
4. Also added checks to make sure the services, interfaces and bgp are up before proceding with the ixia traffic.

### Type of change

- [X] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)

### Back port request
- [ ] 202012
- [ ] 202205
- [ ] 202305
- [ ] 202311
- [X] 202405

#### What is the motivation for this PR?
The issue: https://github.com/sonic-net/sonic-buildimage/issues/20637
#### How did you do it?
Pls see the description.
#### How did you verify/test it?
Ran it on my TB.
```
=========================================================================================================================== PASSES ===========================================================================================================================
____________________________________________________________________________________ test_pfcwd_basic_multi_lossless_prio_restart_service[multidut_port_info0-True-swss] _____________________________________________________________________________________
____________________________________________________________________________________ test_pfcwd_basic_multi_lossless_prio_restart_service[multidut_port_info0-False-swss] ____________________________________________________________________________________
____________________________________________________________________________________ test_pfcwd_basic_multi_lossless_prio_restart_service[multidut_port_info1-True-swss] _____________________________________________________________________________________
____________________________________________________________________________________ test_pfcwd_basic_multi_lossless_prio_restart_service[multidut_port_info1-False-swss] ____________________________________________________________________________________
----------------------------------------------------------------------------- generated xml file: /run_logs/ixia/restart-service/2024-11-14-00-05-11/tr_2024-11-14-00-05-11.xml ------------------------------------------------------------------------------
INFO:root:Can not get Allure report URL. Please check logs
------------------------------------------------------------------------------------------------------------------- live log sessionfinish -------------------------------------------------------------------------------------------------------------------
01:31:34 __init__.pytest_terminal_summary         L0067 INFO   | Can not get Allure report URL. Please check logs
================================================================================================================== short test summary info ===================================================================================================================
PASSED snappi_tests/multidut/pfcwd/test_multidut_pfcwd_basic_with_snappi.py::test_pfcwd_basic_multi_lossless_prio_restart_service[multidut_port_info0-True-swss]
PASSED snappi_tests/multidut/pfcwd/test_multidut_pfcwd_basic_with_snappi.py::test_pfcwd_basic_multi_lossless_prio_restart_service[multidut_port_info0-False-swss]
PASSED snappi_tests/multidut/pfcwd/test_multidut_pfcwd_basic_with_snappi.py::test_pfcwd_basic_multi_lossless_prio_restart_service[multidut_port_info1-True-swss]
PASSED snappi_tests/multidut/pfcwd/test_multidut_pfcwd_basic_with_snappi.py::test_pfcwd_basic_multi_lossless_prio_restart_service[multidut_port_info1-False-swss]
========================================================================================================= 4 passed, 7 warnings in 5180.68s (1:26:20) =========================================================================================================
sonic@ixia-sonic-mgmt-whitebox:/data/tests$

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=========================================================================================================================== PASSES ===========================================================================================================================
____________________________________________________________________________________ test_pfcwd_basic_multi_lossless_prio_restart_service[multidut_port_info0-True-swss] _____________________________________________________________________________________
____________________________________________________________________________________ test_pfcwd_basic_multi_lossless_prio_restart_service[multidut_port_info0-False-swss] ____________________________________________________________________________________
____________________________________________________________________________________ test_pfcwd_basic_multi_lossless_prio_restart_service[multidut_port_info1-True-swss] _____________________________________________________________________________________
____________________________________________________________________________________ test_pfcwd_basic_multi_lossless_prio_restart_service[multidut_port_info1-False-swss] ____________________________________________________________________________________
---------------------------------------------------------------------------- generated xml file: /run_logs/ixia/restart-service-2/2024-11-14-02-47-47/tr_2024-11-14-02-47-47.xml -----------------------------------------------------------------------------
INFO:root:Can not get Allure report URL. Please check logs
------------------------------------------------------------------------------------------------------------------- live log sessionfinish -------------------------------------------------------------------------------------------------------------------
04:14:03 __init__.pytest_terminal_summary         L0067 INFO   | Can not get Allure report URL. Please check logs
================================================================================================================== short test summary info ===================================================================================================================
PASSED snappi_tests/multidut/pfcwd/test_multidut_pfcwd_basic_with_snappi.py::test_pfcwd_basic_multi_lossless_prio_restart_service[multidut_port_info0-True-swss]
PASSED snappi_tests/multidut/pfcwd/test_multidut_pfcwd_basic_with_snappi.py::test_pfcwd_basic_multi_lossless_prio_restart_service[multidut_port_info0-False-swss]
PASSED snappi_tests/multidut/pfcwd/test_multidut_pfcwd_basic_with_snappi.py::test_pfcwd_basic_multi_lossless_prio_restart_service[multidut_port_info1-True-swss]
PASSED snappi_tests/multidut/pfcwd/test_multidut_pfcwd_basic_with_snappi.py::test_pfcwd_basic_multi_lossless_prio_restart_service[multidut_port_info1-False-swss]
========================================================================================================= 4 passed, 7 warnings in 5173.22s (1:26:13) =========================================================================================================
sonic@ixia-sonic-mgmt-whitebox:/data/tests$
=========================================================================================================================== PASSES ===========================================================================================================================
____________________________________________________________________________________ test_pfcwd_basic_single_lossless_prio_service_restart[multidut_port_info0-True-swss] ____________________________________________________________________________________
___________________________________________________________________________________ test_pfcwd_basic_single_lossless_prio_service_restart[multidut_port_info0-False-swss] ____________________________________________________________________________________
____________________________________________________________________________________ test_pfcwd_basic_single_lossless_prio_service_restart[multidut_port_info1-True-swss] ____________________________________________________________________________________
___________________________________________________________________________________ test_pfcwd_basic_single_lossless_prio_service_restart[multidut_port_info1-False-swss] ____________________________________________________________________________________
---------------------------------------------------------------------------- generated xml file: /run_logs/ixia/restart-service-2/2024-11-14-06-39-15/tr_2024-11-14-06-39-15.xml -----------------------------------------------------------------------------
INFO:root:Can not get Allure report URL. Please check logs
------------------------------------------------------------------------------------------------------------------- live log sessionfinish -------------------------------------------------------------------------------------------------------------------
08:10:42 __init__.pytest_terminal_summary         L0067 INFO   | Can not get Allure report URL. Please check logs
================================================================================================================== short test summary info ===================================================================================================================
PASSED snappi_tests/multidut/pfcwd/test_multidut_pfcwd_basic_with_snappi.py::test_pfcwd_basic_single_lossless_prio_service_restart[multidut_port_info0-True-swss]
PASSED snappi_tests/multidut/pfcwd/test_multidut_pfcwd_basic_with_snappi.py::test_pfcwd_basic_single_lossless_prio_service_restart[multidut_port_info0-False-swss]
PASSED snappi_tests/multidut/pfcwd/test_multidut_pfcwd_basic_with_snappi.py::test_pfcwd_basic_single_lossless_prio_service_restart[multidut_port_info1-True-swss]
PASSED snappi_tests/multidut/pfcwd/test_multidut_pfcwd_basic_with_snappi.py::test_pfcwd_basic_single_lossless_prio_service_restart[multidut_port_info1-False-swss]
========================================================================================================= 4 passed, 7 warnings in 5484.86s (1:31:24) =========================================================================================================
sonic@ixia-sonic-mgmt-whitebox:/data/tests$ 
```

### CODE REVIEW

**Comment by rraghav-cisco** [comment]
@sdszhang , @amitpawar12 , @selldinesh : FYI.

File: tests/snappi_tests/multidut/pfcwd/test_multidut_pfcwd_basic_with_snappi.py [hunk 1]
```diff
@@ -264,6 +288,12 @@ def test_pfcwd_basic_single_lossless_prio_service_restart(snappi_api,
             logger.info("Wait until the system is stable")
             pytest_assert(wait_until(WAIT_TIME, INTERVAL, 0, duthost.critical_services_fully_started),
                           "Not all critical services are fully started")
+            pytest_assert(wait_until(WAIT_TIME, INTERVAL, 0, check_interface_status_of_up_ports, duthost),
+                          "Not all interfaces are up.")
+
+        # Wait for internal bgp to come up.
+        time.sleep(100)
```
Comment (sdszhang) [inline tests/snappi_tests/multidut/pfcwd/test_multidut_pfcwd_basic_with_snappi.py:295]: can you change this hard code value to use wait_util()?

here is an example:
https://github.com/sonic-net/sonic-mgmt/blob/72890a4a4e475a255b263a746030bc851426ae21/tests/bgp/test_bgp_session.py#L87

Comment (rraghav-cisco) [reply tests/snappi_tests/multidut/pfcwd/test_multidut_pfcwd_basic_with_snappi.py:295]: Thanks for the examples. I am not sure if they work with some of the BGP sessions down. In our case, the backplane BGP are supposed to come up, but the front-facing(or ixia facing) BGP sessions are not brought up. I have to check if the above function can take only the backplane bgp neighbors.

Comment (sdszhang) [reply tests/snappi_tests/multidut/pfcwd/test_multidut_pfcwd_basic_with_snappi.py:295]: @rraghav-cisco you can use 'INTERNAL' to filter only iBGP sessions.

https://github.com/sonic-net/sonic-mgmt/blob/eb0081706b5166d4fa36255e7eb9504eedb32901/tests/bgp/test_bgp_session_flap.py#L70

File: tests/snappi_tests/multidut/pfcwd/test_multidut_pfcwd_basic_with_snappi.py [hunk 2]
```diff
@@ -361,6 +361,7 @@ def test_pfcwd_basic_multi_lossless_prio_restart_service(snappi_api,
 
         logger.info('Port dictionary:{}'.format(ports_dict))
         for duthost in list(set([snappi_ports[0]['duthost'], snappi_ports[1]['duthost']])):
+            up_bgp_neighbors = duthost.get_bgp_neighbors_per_asic("established")
```
Comment (amitpawar12) [inline tests/snappi_tests/multidut/pfcwd/test_multidut_pfcwd_basic_with_snappi.py:364]: @rraghav-cisco  - the BGP state check is ONLY if the duthost is multi.asic. Is that not true for other systems?

File: tests/snappi_tests/multidut/pfcwd/test_multidut_pfcwd_basic_with_snappi.py (hunk 1 above)
Comment (rraghav-cisco) [reply tests/snappi_tests/multidut/pfcwd/test_multidut_pfcwd_basic_with_snappi.py]: @sdszhang : Done.

File: tests/snappi_tests/multidut/pfcwd/test_multidut_pfcwd_basic_with_snappi.py (hunk 2 above)
Comment (rraghav-cisco) [reply tests/snappi_tests/multidut/pfcwd/test_multidut_pfcwd_basic_with_snappi.py:364]: I am focusing only on T2/multiasic at the moment. The non-mAsic systems are in the else block. I will have to check that when t0/t1/t2 are all combined.

Comment (rraghav-cisco) [reply tests/snappi_tests/multidut/pfcwd/test_multidut_pfcwd_basic_with_snappi.py:364]: Also, the internal BGP routes are not present for T0/T1. Only T2 has them. The other checks - services and interfaces will be needed for T0/T1 as well.

Comment (sdszhang) [reply tests/snappi_tests/multidut/pfcwd/test_multidut_pfcwd_basic_with_snappi.py:364]: as discussed, @rraghav-cisco pls open a separate PR to handle T0/T1 case.

**Pull request merged**